File: core/whitemagic/inference/routing_metrics.py
# ruff: noqa: BLE001
"""Routing Observability — tracks inference routing decisions, latency, and escalation rates.

Production hybrid routing systems require telemetry to detect threshold drift,
measure per-tier quality, and tune routing parameters. This module provides
rolling-window statistics for the InferenceRouter.

Tracks:
  - Per-tier request counts and latency (p50, p95, p99)
  - Escalation rates (Tier N → Tier N+1)
  - Confidence distribution at each tier
  - Routing decision reasons
  - Circuit breaker state per tier
"""

from __future__ import annotations

import logging
import threading
import time
from collections import defaultdict, deque
from dataclasses import dataclass, field
from typing import Any

from whitemagic.inference.complexity import InferenceTier

logger = logging.getLogger(__name__)

_WINDOW_SIZE = 1000  # Rolling window for latency samples


@dataclass
class TierStats:
    """Statistics for a single inference tier."""

    total_requests: int = 0
    successful: int = 0
    failed: int = 0
    escalated: int = 0  # Requests that escalated to a higher tier
    latencies: deque = field(default_factory=lambda: deque(maxlen=_WINDOW_SIZE))
    confidences: deque = field(default_factory=lambda: deque(maxlen=_WINDOW_SIZE))
    decision_reasons: dict[str, int] = field(default_factory=lambda: defaultdict(int))

    @property
    def success_rate(self) -> float:
        if self.total_requests == 0:
            return 0.0
        return self.successful / self.total_requests

    @property
    def escalation_rate(self) -> float:
        if self.total_requests == 0:
            return 0.0
        return self.escalated / self.total_requests

    def record(
        self, latency_ms: float, confidence: float, success: bool, reason: str = ""
    ) -> None:
        self.total_requests += 1
        self.latencies.append(latency_ms)
        self.confidences.append(confidence)
        if success:
            self.successful += 1
        else:
            self.failed += 1
        if reason:
            self.decision_reasons[reason] += 1

    def record_escalation(self) -> None:
        self.escalated += 1

    def p50(self) -> float:
        return self._percentile(50)

    def p95(self) -> float:
        return self._percentile(95)

    def p99(self) -> float:
        return self._percentile(99)

    def avg_confidence(self) -> float:
        if not self.confidences:
            return 0.0
        return sum(self.confidences) / len(self.confidences)

    def _percentile(self, p: int) -> float:
        if not self.latencies:
            return 0.0
        sorted_lat = sorted(self.latencies)
        idx = int(len(sorted_lat) * p / 100)
        return float(sorted_lat[min(idx, len(sorted_lat) - 1)])

    def to_dict(self) -> dict[str, Any]:
        return {
            "total_requests": self.total_requests,
            "successful": self.successful,
            "failed": self.failed,
            "escalated": self.escalated,
            "success_rate": f"{self.success_rate:.1%}",
            "escalation_rate": f"{self.escalation_rate:.1%}",
            "p50_ms": round(self.p50(), 2),
            "p95_ms": round(self.p95(), 2),
            "p99_ms": round(self.p99(), 2),
            "avg_confidence": round(self.avg_confidence(), 3),
            "top_reasons": dict(
                sorted(self.decision_reasons.items(), key=lambda x: -x[1])[:5]
            ),
        }


class RoutingMetrics:
    """Rolling-window metrics for inference routing decisions.

    Thread-safe. Designed for the routing hot path — all operations are O(1)
    except percentile calculations which are O(n log n) on a bounded window.
    """

    def __init__(self) -> None:
        self._tier_stats: dict[InferenceTier, TierStats] = {
            tier: TierStats() for tier in InferenceTier
        }
        self._total_routed: int = 0
        self._total_escalations: int = 0
        self._lock = threading.RLock()
        self._start_time = time.time()

    def record_routing(
        self,
        tier: InferenceTier,
        latency_ms: float,
        confidence: float,
        success: bool,
        reason: str = "",
    ) -> None:
        """Record a routing decision and its outcome."""
        with self._lock:
            self._total_routed += 1
            self._tier_stats[tier].record(latency_ms, confidence, success, reason)

    def record_escalation(
        self,
        from_tier: InferenceTier,
        to_tier: InferenceTier,
        reason: str = "",
    ) -> None:
        """Record an escalation from one tier to a higher one."""
        with self._lock:
            self._total_escalations += 1
            self._tier_stats[from_tier].record_escalation()
            logger.debug(
                "Escalation %s → %s (reason: %s). Total escalations: %d",
                from_tier.name,
                to_tier.name,
                reason,
                self._total_escalations,
            )

    def summary(self) -> dict[str, Any]:
        """Get a summary of all routing metrics."""
        with self._lock:
            uptime = time.time() - self._start_time
            return {
                "uptime_seconds": round(uptime, 1),
                "total_routed": self._total_routed,
                "total_escalations": self._total_escalations,
                "overall_escalation_rate": (
                    f"{self._total_escalations / max(1, self._total_routed):.1%}"
                ),
                "tiers": {
                    tier.name: stats.to_dict()
                    for tier, stats in self._tier_stats.items()
                    if stats.total_requests > 0
                },
            }

    def detect_drift(self, window: int = 100) -> dict[str, Any]:
        """Detect routing threshold drift.

        Checks if escalation rates have shifted significantly from the
        historical baseline, which may indicate the routing thresholds
        need retuning.

        Args:
            window: Number of recent samples to compare against full history.

        Returns:
            Drift analysis with recommendations.
        """
        with self._lock:
            drift: dict[str, Any] = {"status": "ok", "recommendations": []}

            for tier, stats in self._tier_stats.items():
                if stats.total_requests < window * 2:
                    continue

                # Compare recent escalation rate to historical
                recent_escalations = sum(
                    1
                    for i in range(min(window, len(stats.latencies)))
                    if i < stats.escalated  # Approximate
                )
                recent_rate = recent_escalations / min(window, stats.total_requests)
                historical_rate = stats.escalation_rate

                if historical_rate > 0 and abs(recent_rate - historical_rate) > 0.15:
                    drift["status"] = "drift_detected"
                    drift["recommendations"].append(
                        {
                            "tier": tier.name,
                            "historical_rate": f"{historical_rate:.1%}",
                            "recent_rate": f"{recent_rate:.1%}",
                            "action": "Consider adjusting confidence threshold or complexity patterns",
                        }
                    )

            return drift

    def adaptive_thresholds(self) -> dict[str, float]:
        """Compute adaptive confidence thresholds based on observed performance.

        If a tier has high success rate and low escalation, lower the threshold
        (keep more requests at that tier). If escalation rate is high, raise the
        threshold (send fewer requests to that tier).

        Returns:
            Dict mapping tier name to recommended confidence threshold.
        """
        with self._lock:
            recommendations: dict[str, float] = {}
            base_threshold = 0.85

            for tier, stats in self._tier_stats.items():
                if stats.total_requests < 20:
                    recommendations[tier.name] = base_threshold
                    continue

                # High escalation rate → raise threshold (be more conservative)
                # Low escalation rate + high success → lower threshold (be more aggressive)
                esc_rate = stats.escalation_rate
                succ_rate = stats.success_rate

                # Adjust threshold: ±0.05 max deviation
                adjustment = 0.0
                if esc_rate > 0.3:
                    adjustment = 0.05  # Raise threshold
                elif esc_rate < 0.05 and succ_rate > 0.9:
                    adjustment = -0.05  # Lower threshold

                recommended = max(0.5, min(0.99, base_threshold + adjustment))
                recommendations[tier.name] = round(recommended, 3)

            return recommendations

    def export_dashboard(self) -> dict[str, Any]:
        """Export metrics in a format suitable for the Mound/dashboard.

        Returns a flat dict with key metrics for monitoring:
        - Per-tier counts, rates, latencies
        - Overall escalation rate
        - Drift status
        - Adaptive threshold recommendations
        """
        with self._lock:
            summary = self.summary()
            drift = self.detect_drift()
            thresholds = self.adaptive_thresholds()

            return {
                "routing": summary,
                "drift": drift,
                "adaptive_thresholds": thresholds,
                "exported_at": time.time(),
            }

    def to_mound_metrics(self) -> dict[str, float]:
        """Export key metrics as flat float dict for Mound integration.

        The Mound Gana tracks numeric metrics. This method flattens
        routing observability into individual metric keys.
        """
        with self._lock:
            metrics: dict[str, float] = {
                "routing.total_routed": float(self._total_routed),
                "routing.total_escalations": float(self._total_escalations),
                "routing.uptime_seconds": time.time() - self._start_time,
            }

            for tier, stats in self._tier_stats.items():
                prefix = f"routing.{tier.name.lower()}"
                metrics[f"{prefix}.requests"] = float(stats.total_requests)
                metrics[f"{prefix}.success_rate"] = stats.success_rate
                metrics[f"{prefix}.escalation_rate"] = stats.escalation_rate
                metrics[f"{prefix}.p50_ms"] = stats.p50()
                metrics[f"{prefix}.p95_ms"] = stats.p95()
                metrics[f"{prefix}.p99_ms"] = stats.p99()
                metrics[f"{prefix}.avg_confidence"] = stats.avg_confidence()

            return metrics


# Singleton
_metrics: RoutingMetrics | None = None


def get_routing_metrics() -> RoutingMetrics:
    """Get singleton routing metrics."""
    global _metrics
    if _metrics is None:
        _metrics = RoutingMetrics()
    return _metrics

File: crates/wm-bicameral/src/routing_metrics.rs
//! Routing Observability — rolling-window metrics for inference routing.
//!
//! Ported from v2's inference/routing_metrics.py.
//! Tracks per-tier request counts, latency percentiles (p50/p95/p99),
//! escalation rates, confidence distributions, and routing decision reasons.
//! Provides drift detection and adaptive threshold recommendations.

use std::collections::{HashMap, VecDeque};
use std::time::Instant;

use crate::router::InferenceTier;

const WINDOW_SIZE: usize = 1000;

/// Statistics for a single inference tier.
#[derive(Debug)]
pub struct TierStats {
    pub total_requests: u64,
    pub successful: u64,
    pub failed: u64,
    pub escalated: u64,
    latencies: VecDeque<f64>,
    confidences: VecDeque<f64>,
    decision_reasons: HashMap<String, u64>,
}

impl TierStats {
    /// Create empty tier stats.
    #[must_use]
    pub fn new() -> Self {
        Self {
            total_requests: 0,
            successful: 0,
            failed: 0,
            escalated: 0,
            latencies: VecDeque::with_capacity(WINDOW_SIZE),
            confidences: VecDeque::with_capacity(WINDOW_SIZE),
            decision_reasons: HashMap::new(),
        }
    }

    /// Success rate (0.0–1.0).
    #[must_use]
    pub fn success_rate(&self) -> f64 {
        if self.total_requests == 0 {
            0.0
        } else {
            self.successful as f64 / self.total_requests as f64
        }
    }

    /// Escalation rate (0.0–1.0).
    #[must_use]
    pub fn escalation_rate(&self) -> f64 {
        if self.total_requests == 0 {
            0.0
        } else {
            self.escalated as f64 / self.total_requests as f64
        }
    }

    /// Record a routing outcome.
    pub fn record(&mut self, latency_ms: f64, confidence: f64, success: bool, reason: &str) {
        self.total_requests += 1;
        if self.latencies.len() >= WINDOW_SIZE {
            self.latencies.pop_front();
        }
        self.latencies.push_back(latency_ms);
        if self.confidences.len() >= WINDOW_SIZE {
            self.confidences.pop_front();
        }
        self.confidences.push_back(confidence);
        if success {
            self.successful += 1;
        } else {
            self.failed += 1;
        }
        if !reason.is_empty() {
            *self.decision_reasons.entry(reason.to_string()).or_insert(0) += 1;
        }
    }

    /// Record an escalation from this tier.
    pub const fn record_escalation(&mut self) {
        self.escalated += 1;
    }

    /// 50th percentile latency in ms.
    #[must_use]
    pub fn p50(&self) -> f64 {
        self.percentile(50)
    }

    /// 95th percentile latency in ms.
    #[must_use]
    pub fn p95(&self) -> f64 {
        self.percentile(95)
    }

    /// 99th percentile latency in ms.
    #[must_use]
    pub fn p99(&self) -> f64 {
        self.percentile(99)
    }

    /// Average confidence (0.0–1.0).
    #[must_use]
    pub fn avg_confidence(&self) -> f64 {
        if self.confidences.is_empty() {
            0.0
        } else {
            self.confidences.iter().sum::<f64>() / self.confidences.len() as f64
        }
    }

    fn percentile(&self, p: u32) -> f64 {
        if self.latencies.is_empty() {
            return 0.0;
        }
        let mut sorted: Vec<f64> = self.latencies.iter().copied().collect();
        sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
        let idx = (sorted.len() * p as usize) / 100;
        sorted[idx.min(sorted.len() - 1)]
    }

    /// Top N decision reasons by count.
    #[must_use]
    pub fn top_reasons(&self, n: usize) -> Vec<(String, u64)> {
        let mut reasons: Vec<(String, u64)> = self
            .decision_reasons
            .iter()
            .map(|(k, &v)| (k.clone(), v))
            .collect();
        reasons.sort_by_key(|x| std::cmp::Reverse(x.1));
        reasons.truncate(n);
        reasons
    }
}

impl Default for TierStats {
    fn default() -> Self {
        Self::new()
    }
}

/// Rolling-window metrics for inference routing decisions.
///
/// All operations are O(1) except percentile calculations which are
/// O(n log n) on a bounded window (max 1000 samples).
pub struct RoutingMetrics {
    tier_stats: HashMap<InferenceTier, TierStats>,
    total_routed: u64,
    total_escalations: u64,
    start_time: Instant,
}

impl Default for RoutingMetrics {
    fn default() -> Self {
        Self::new()
    }
}

impl RoutingMetrics {
    /// Create a new routing metrics collector.
    #[must_use]
    pub fn new() -> Self {
        let mut tier_stats = HashMap::new();
        for tier in InferenceTier::all() {
            tier_stats.insert(tier, TierStats::new());
        }
        Self {
            tier_stats,
            total_routed: 0,
            total_escalations: 0,
            start_time: Instant::now(),
        }
    }

    /// Record a routing decision and its outcome.
    pub fn record_routing(
        &mut self,
        tier: InferenceTier,
        latency_ms: f64,
        confidence: f64,
        success: bool,
        reason: &str,
    ) {
        self.total_routed += 1;
        if let Some(stats) = self.tier_stats.get_mut(&tier) {
            stats.record(latency_ms, confidence, success, reason);
        }
    }

    /// Record an escalation from one tier to a higher one.
    pub fn record_escalation(
        &mut self,
        from_tier: InferenceTier,
        _to_tier: InferenceTier,
        _reason: &str,
    ) {
        self.total_escalations += 1;
        if let Some(stats) = self.tier_stats.get_mut(&from_tier) {
            stats.record_escalation();
        }
    }

    /// Get stats for a specific tier.
    #[must_use]
    pub fn tier_stats(&self, tier: InferenceTier) -> Option<&TierStats> {
        self.tier_stats.get(&tier)
    }

    /// Total number of routed requests.
    #[must_use]
    pub const fn total_routed(&self) -> u64 {
        self.total_routed
    }

    /// Total number of escalations.
    #[must_use]
    pub const fn total_escalations(&self) -> u64 {
        self.total_escalations
    }

    /// Overall escalation rate (0.0–1.0).
    #[must_use]
    pub fn overall_escalation_rate(&self) -> f64 {
        if self.total_routed == 0 {
            0.0
        } else {
            self.total_escalations as f64 / self.total_routed as f64
        }
    }

    /// Uptime in seconds.
    #[must_use]
    pub fn uptime_seconds(&self) -> f64 {
        self.start_time.elapsed().as_secs_f64()
    }

    /// Detect routing threshold drift.
    ///
    /// Checks if escalation rates have shifted significantly from the
    /// historical baseline, which may indicate routing thresholds need
    /// retuning.
    #[must_use]
    pub fn detect_drift(&self, window: u64) -> DriftReport {
        let mut recommendations = Vec::new();

        for tier in InferenceTier::all() {
            if let Some(stats) = self.tier_stats.get(&tier) {
                if stats.total_requests < window * 2 {
                    continue;
                }

                let historical_rate = stats.escalation_rate();
                // Approximate recent rate: compare escalated in recent window
                let recent_escalations = stats.escalated.min(window);
                let recent_rate = recent_escalations as f64 / window as f64;

                if historical_rate > 0.0 && (recent_rate - historical_rate).abs() > 0.15 {
                    recommendations.push(DriftRecommendation {
                        tier,
                        historical_rate,
                        recent_rate,
                    });
                }
            }
        }

        let status = if recommendations.is_empty() {
            DriftStatus::Ok
        } else {
            DriftStatus::DriftDetected
        };

        DriftReport {
            status,
            recommendations,
        }
    }

    /// Compute adaptive confidence thresholds based on observed performance.
    ///
    /// If a tier has high success rate and low escalation, lower the threshold
    /// (keep more requests at that tier). If escalation rate is high, raise the
    /// threshold (send fewer requests to that tier).
    #[must_use]
    pub fn adaptive_thresholds(&self) -> HashMap<InferenceTier, f32> {
        let base_threshold = 0.85_f32;
        let mut result = HashMap::new();

        for tier in InferenceTier::all() {
            if let Some(stats) = self.tier_stats.get(&tier) {
                if stats.total_requests < 20 {
                    result.insert(tier, base_threshold);
                    continue;
                }

                let esc_rate = stats.escalation_rate();
                let succ_rate = stats.success_rate();

                let adjustment = if esc_rate > 0.3 {
                    0.05 // Raise threshold
                } else if esc_rate < 0.05 && succ_rate > 0.9 {
                    -0.05 // Lower threshold
                } else {
                    0.0
                };

                let recommended = (base_threshold + adjustment).clamp(0.5, 0.99);
                result.insert(tier, recommended);
            } else {
                result.insert(tier, base_threshold);
            }
        }

        result
    }
}

/// Drift detection status.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DriftStatus {
    /// No drift detected
    Ok,
    /// Drift detected — thresholds may need retuning
    DriftDetected,
}

/// A single drift recommendation.
#[derive(Debug, Clone)]
pub struct DriftRecommendation {
    /// Tier with detected drift
    pub tier: InferenceTier,
    /// Historical escalation rate
    pub historical_rate: f64,
    /// Recent escalation rate
    pub recent_rate: f64,
}

/// Result of drift detection.
#[derive(Debug, Clone)]
pub struct DriftReport {
    /// Overall drift status
    pub status: DriftStatus,
    /// Per-tier drift recommendations
    pub recommendations: Vec<DriftRecommendation>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tier_stats_empty() {
        let stats = TierStats::new();
        assert_eq!(stats.total_requests, 0);
        assert_eq!(stats.success_rate(), 0.0);
        assert_eq!(stats.escalation_rate(), 0.0);
        assert_eq!(stats.p50(), 0.0);
        assert_eq!(stats.avg_confidence(), 0.0);
    }

    #[test]
    fn tier_stats_record() {
        let mut stats = TierStats::new();
        stats.record(50.0, 0.9, true, "low_complexity");
        stats.record(100.0, 0.7, true, "low_complexity");
        stats.record(200.0, 0.5, false, "timeout");

        assert_eq!(stats.total_requests, 3);
        assert_eq!(stats.successful, 2);
        assert_eq!(stats.failed, 1);
        assert!((stats.success_rate() - 2.0 / 3.0).abs() < 0.01);
    }

    #[test]
    fn tier_stats_percentiles() {
        let mut stats = TierStats::new();
        for i in 1..=100 {
            stats.record(f64::from(i), 0.8, true, "");
        }
        assert!(stats.p50() > 0.0);
        assert!(stats.p95() >= stats.p50());
        assert!(stats.p99() >= stats.p95());
    }

    #[test]
    fn tier_stats_escalation() {
        let mut stats = TierStats::new();
        stats.record(50.0, 0.9, true, "");
        stats.record(100.0, 0.5, true, "");
        stats.record_escalation();

        assert_eq!(stats.escalated, 1);
        assert!((stats.escalation_rate() - 0.5).abs() < 0.01);
    }

    #[test]
    fn tier_stats_top_reasons() {
        let mut stats = TierStats::new();
        stats.record(50.0, 0.9, true, "fast");
        stats.record(50.0, 0.9, true, "fast");
        stats.record(50.0, 0.9, true, "slow");

        let top = stats.top_reasons(5);
        assert_eq!(top.len(), 2);
        assert_eq!(top[0].0, "fast");
        assert_eq!(top[0].1, 2);
    }

    #[test]
    fn routing_metrics_record() {
        let mut metrics = RoutingMetrics::new();
        metrics.record_routing(InferenceTier::EdgeRules, 0.5, 0.95, true, "pattern_match");
        metrics.record_routing(InferenceTier::LocalSmall, 100.0, 0.8, true, "moderate");

        assert_eq!(metrics.total_routed(), 2);
        assert_eq!(metrics.total_escalations(), 0);
    }

    #[test]
    fn routing_metrics_escalation() {
        let mut metrics = RoutingMetrics::new();
        metrics.record_routing(InferenceTier::EdgeRules, 0.5, 0.3, false, "low_confidence");
        metrics.record_escalation(
            InferenceTier::EdgeRules,
            InferenceTier::LocalSmall,
            "low_confidence",
        );

        assert_eq!(metrics.total_escalations(), 1);
        assert!((metrics.overall_escalation_rate() - 1.0).abs() < f64::from(f32::EPSILON));
    }

    #[test]
    fn routing_metrics_adaptive_thresholds() {
        let mut metrics = RoutingMetrics::new();

        // Record many successful low-escalation requests
        for _ in 0..30 {
            metrics.record_routing(InferenceTier::EdgeRules, 0.5, 0.95, true, "");
        }

        let thresholds = metrics.adaptive_thresholds();
        let edge_threshold = thresholds[&InferenceTier::EdgeRules];
        // High success, low escalation → threshold should be lowered
        assert!(edge_threshold < 0.85);
    }

    #[test]
    fn routing_metrics_adaptive_thresholds_high_escalation() {
        let mut metrics = RoutingMetrics::new();

        // Record requests with high escalation
        for _ in 0..30 {
            metrics.record_routing(InferenceTier::LocalSmall, 100.0, 0.5, true, "");
            metrics.record_escalation(InferenceTier::LocalSmall, InferenceTier::LocalLarge, "");
        }

        let thresholds = metrics.adaptive_thresholds();
        let threshold = thresholds[&InferenceTier::LocalSmall];
        // High escalation → threshold should be raised
        assert!(threshold > 0.85);
    }

    #[test]
    fn routing_metrics_drift_ok() {
        let mut metrics = RoutingMetrics::new();
        for _ in 0..200 {
            metrics.record_routing(InferenceTier::EdgeRules, 0.5, 0.9, true, "");
        }
        let report = metrics.detect_drift(100);
        assert_eq!(report.status, DriftStatus::Ok);
    }

    #[test]
    fn routing_metrics_tier_stats_access() {
        let mut metrics = RoutingMetrics::new();
        metrics.record_routing(InferenceTier::Cloud, 5000.0, 0.99, true, "complex");
        let stats = metrics.tier_stats(InferenceTier::Cloud).unwrap();
        assert_eq!(stats.total_requests, 1);
    }
}
